### PR TITLE
Try to connect to all IP addresses simultaneously

### PR DESCRIPTION
**Pull request by duderamos**
* In case a database service is a hostname and DNS returns multiple
addresses, try to connect to all in parallel. It will fail faster and
get a possible valid connection quicker.

### CODE REVIEW

File: lib/active_record/connection_adapters/sqlserver_adapter.rb [hunk 1]
```diff
@@ -315,37 +314,55 @@ def connection_errors
...
-          else
-            client.execute('SET ANSI_DEFAULTS ON').do
-            client.execute('SET CURSOR_CLOSE_ON_COMMIT OFF').do
-            client.execute('SET IMPLICIT_TRANSACTIONS OFF').do
+        hosts = config_hosts(config)
+        threads = []
+        client = nil
+
+        hosts.each do |host|
+          threads << Thread.new do
```
Comment (martinbjeldbak) [inline lib/active_record/connection_adapters/sqlserver_adapter.rb]: Just a thought - should we consider `concurrent-ruby` here or is using `Thread` enough?

File: lib/active_record/connection_adapters/sqlserver_adapter.rb [hunk 2]
```diff
@@ -383,6 +400,14 @@ def config_encoding(config)
         config[:encoding].present? ? config[:encoding] : nil
       end
 
+      def config_hosts(config)
+        resolver = Resolv::DNS.open
+        hosts = resolver.getaddresses(config[:host]).map(&:to_s)
```
Comment (martinbjeldbak) [inline lib/active_record/connection_adapters/sqlserver_adapter.rb]: what if `config[:host]` is an IP address? does this transparently work still?

File: lib/active_record/connection_adapters/sqlserver_adapter.rb (hunk 1 above)
Comment (duderamos) [reply lib/active_record/connection_adapters/sqlserver_adapter.rb]: I think `Thread` is enough for now, and it does not require a new dependency.

File: lib/active_record/connection_adapters/sqlserver_adapter.rb (hunk 2 above)
Comment (duderamos) [reply lib/active_record/connection_adapters/sqlserver_adapter.rb]: That's a good point. I expected that it was smart enough to return the IP in that case. I was wrong.

**Comment by duderamos** [comment]
There're more things to do before it could be taken seriously. I think the best would be having a config param like `failover_mode` to enable or disable this `special` behaviour. For now, it just works with me very specific scenario/configuration.

**Pull request closed without merging**